### ISSUE TITLE
OpenMP Support with Clang (`clang-cl`) on Windows

### ISSUE THREAD

**Issue by RoyiAvital** [mesonbuild/meson#5298]
Hello,

When using Clang compiler on Windows (`clang-cl`) the `dependency(openmp)` doesn't work.
What should happen is looking for `libomp.lib` in the `lib` folder in the installation folder of `clang-cl`.

The problem is even harder as there is a bug when trying to pass the compiler the flags: `-Xclang -fopenmp`.
They are surrounded with `"` in the `build.ninja` file. For some reason this makes Clang ignore it.

What I could do manually is change:

```
ARGS = "-IMyFile@exe" "-I." "-I.." "/MD" "/nologo" "/showIncludes" "/W4" "/Ob2" "/Gw" "/O2" "-Xclang$ -fopenmp" "/arch:AVX2" "/IMyFile@exe\\IMyFile_IMyFile.c.pdb"
```
Into:
```
ARGS = "-IMyFile@exe" "-I." "-I.." "/MD" "/nologo" "/showIncludes" "/W4" "/Ob2" "/Gw" "/O2" -Xclang -fopenmp "/arch:AVX2" "/IMyFile@exe\\IMyFile_IMyFile.c.pdb"
```

Which make it work.
Namely the `"` should be removed.

Could we do something to make those appear in a correct form to begin with?

**Comment by RoyiAvital** [comment]
OK,
I think I found another solution.

It seems that `clang-cl` parse anything which comes after `-Xclang` as `clang` parameter (And not only single one as I was told).

Hence it is important to keep the native `clang` parameter at the end with no `MSVC` parameters after them.

So something like: `compFlags = ['/arch:AVX2', '-Xclang', '-fopenmp']` will work when parsed.

**Comment by dcbaker** [comment]
Does passing '/openmp' to clang-cl work, or does it need `-Xclang -fopenmp`?

**Comment by RoyiAvital** [comment]
First, I'd expect `dependency(openmp)` to work on `clang-cl`.

Also, both `-openmp` and `/openmp` didn't work (Yielding - `clang-cl.exe: warning: argument unused during compilation: '/openmp' [-Wunused-command-line-argument]`).

By the way, I think that `Release` build on `clang-cl` should have option for `/Ofast` and not only `/O2`.
Or at least we should have easy access to set `-Ofast` instead of `/O2`.

**Comment by dcbaker** [comment]
I sent a PR that I think should fix it (I don't have my windows machine at the moment), so if you could test it that would be helpful.

**Comment by jon-turney** [comment]
Associated clang bug: https://bugs.llvm.org/show_bug.cgi?id=31207

**Comment by RoyiAvital** [comment]
@jon-turney , Quite old bug.
I wish that if they fix it they will give the user 2 options:
 * Link against MSVC's OpenMP library (Very old, Something like OpenMP 2 only).
 * Link against Clang OpenMP library (Mature, based on Intel `libiomp5`) like done with `-fopenmp`.

**Comment by XVilka** [comment]
Note that clang bug was fixed.

**Comment by peter-urban** [comment]
Hello,

I cannot get openmp and clang-cl to run and so I was wondering what the current status of this bug is?

Currently I use
```
omp_dep = dependency('openmp')
```
This works on linux (gcc) and mac (clang). But on windows with clang-cl  I keep receiving 
```lld-link: error: undefined symbol: __kmpc_global_thread_num" ```
and other kmpc linker errors. I tried ( in meson.build)
```
add_global_arguments('-Xclang','-fopenmp', language : 'cpp')
```
But without success.

The test system is (github actions / windows)
- clang-cl 16.0.6
- open mp 5.0
- meson 1.2.1

**Comment by YouFoundJK** [comment]
@peter-urban Did you manage to fix this OpenMP lld-linker issue? If yes could you please explain how you did it? I'm also having the same issue now. Thanks in advance!

**Comment by dcbaker** [comment]
Can you try adding
```meson
omp = meson.get_compiler('cpp').find_library('omp')
 uild_target(…, link_with : omp) # your target using OpenMP
```
a little googling suggests that may be necessary. We can fix that in the implementation if we can verify that works

**Comment by peter-urban** [comment]
@YouFoundJK: No i didn't get this to work yet 

@dcbaker: thanks for your reply and the suggestion, the example code gave me 2 errors:

1) "No library called omp". This could be fixed using omp = meson.get_compiler('cpp').find_library('libomp')

2) "link_with does only work with custom targets". So I added omp to "dependencies"

The following code compiles for me

```
if meson.get_compiler('cpp').get_id() == 'clang-cl'
  libomp = meson.get_compiler('cpp').find_library('libomp')
  omp_dep = declare_dependency(
  compile_args : ['-fopenmp'],
  dependencies : [libomp]
  )
else
  omp_dep = dependency('openmp')
endif

target = library( ..., dependencies=[ omp_dep, ...])
```

However it seems that this does not activate openmp (I do not see multiple processors used on windows).

That said, my project became too complex to debug this issue.  (the error messages for omp_dep = dependency('openmp') on clang-cl changed and became very unspecific.)

I will thus create a dedicated meson / openmp project and post the results once done

**Comment by peter-urban** [comment]
I created a meson openmp test repository here: https://github.com/peter-urban/demonstrate-meson-openmp

The repositories includes three tests to see if openmp is working properly. But I now see that one would have been enough:
- the simple question: is _OPENMP defined

the tests work well for:
- linux (ubuntu) with gcc 
- macos with clang 
- windows with msvc

But I still can't get this to work for windows and clang-cl (https://github.com/peter-urban/demonstrate-meson-openmp/actions/workflows/ci-windows-clang-cl.yml)

I tried with 

```
openmp = dependency('openmp') 
```

(https://github.com/peter-urban/demonstrate-meson-openmp/actions/runs/7465330885)

and with 

```
libomp = meson.get_compiler('cpp').find_library('libomp')
  omp_dep = declare_dependency(
  compile_args : ['-fopenmp'],
  dependencies : [libomp]
  )
```

(https://github.com/peter-urban/demonstrate-meson-openmp/actions/runs/7465338972)

Admittedly I do not know if this is a meson problem, or if I setup clang-cl wrongly in github actions.

Any help,  hints or comments are welcome!

**Comment by peter-urban** [comment]
Ok, I got it :partying_face:  :balloon: 

there are two possible solutions. It is possible to use the flag '/openmp'  (https://github.com/peter-urban/demonstrate-meson-openmp/actions/runs/7467600619/job/20321420466)

```
# openmp
if meson.get_compiler('cpp').get_id() == 'clang-cl'
  libomp = meson.get_compiler('cpp').find_library('libomp')

  openmp = declare_dependency(
    compile_args : ['/openmp'],
    dependencies : [libomp]
  )

  # openmp = dependency('openmp')
else
  openmp = dependency('openmp')
endif
```

or '/clang:-fopenmp': (https://github.com/peter-urban/demonstrate-meson-openmp/actions/runs/7467616389/job/20321464726)

```
# openmp
if meson.get_compiler('cpp').get_id() == 'clang-cl'
  libomp = meson.get_compiler('cpp').find_library('libomp')

  openmp = declare_dependency(
    compile_args : ['/clang:-fopenmp'],
    dependencies : [libomp]
  )

  # openmp = dependency('openmp')
else
  openmp = dependency('openmp')
endif
```

I am happy to finally have a work around. But this should of cause be handled by meson in dependency('openmp') directly.

**Comment by peter-urban** [comment]
Investigating where this has to be implemented: I guess the /openmp flag is the correct one, since this is already implemented for https://github.com/mesonbuild/meson/blob/ce2db13e94fdeaff72d9e1f23b40072f9fb8818b/mesonbuild/compilers/mixins/visualstudio.py#L207 for VisualStudioLikeCompiler ?

**Comment by eli-schwartz** [comment]
clang-cl usually behaves like MSVC, as I understand it, so yes.

**Comment by peter-urban** [comment]
I am working on a PR, but I still have trouble understanding the code base. 

I suppose a good solution is to overwrite the ```openmp_link_flags``` function for the ```ClangCLCompiler``` class (https://github.com/mesonbuild/meson/blob/ce2db13e94fdeaff72d9e1f23b40072f9fb8818b/mesonbuild/compilers/mixins/visualstudio.py#L429) to return an output similar to ```meson.get_compiler('cpp').find_library('libomp')```.

Since ```ClangCLCompiler``` derives from the ```Compile``` class, it implements the function ```self.find_library```. However, here this function also requires me to pass ```env``` and I do not know what to pass here. 

@dcbaker, @eli-schwartz do you have some hints for me?

**Comment by eli-schwartz** [comment]
What you actually want is this:
https://github.com/mesonbuild/meson/blob/ce2db13e94fdeaff72d9e1f23b40072f9fb8818b/mesonbuild/dependencies/misc.py#L74-L131

This does find_library, has_header etc checks and also consumes openmp_flags / openmp_link_flags as needed. The latter functions are only intended to specify the additional compiler-specific flags that should be used *iff* openmp support exists -- they don't check for openmp support.

**Comment by peter-urban** [comment]
Thanks! I can see that the dependencies have a ```selv.env``` variable that I can pass, so I should be able to call ```get_compiler().find_library``` from there

I'll draft a PR for this 👍

Gesendet von Outlook für Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by eli-schwartz** [comment]
Should be fixed by #12723
